### ISSUE TITLE
 Issue with Azure Search Deployment: AuthorizationFailed (403) Error

### ISSUE THREAD

**Issue by Kunal1505** [Azure-Samples/azure-search-openai-demo#582]
I encountered an issue while deploying Azure Search, and the deployment failed with the following error message:
ERROR: deployment failed: failing invoking action 'provision', error deploying infrastructure: starting deployment to subscription: PUT https://management.azure.com/subscriptions/xxxx/providers/Microsoft.Resources/deployments/azure-cog-search-dev-98989
**--------------------------------------------------------------------------------
RESPONSE 403: 403 Forbidden
ERROR CODE: AuthorizationFailed
--------------------------------------------------------------------------------
{
  "error": {
    "code": "AuthorizationFailed",
    "message": "The client 'user@xx.com' with object id 'xxxx' does not have authorization to perform actionve authorization to perform action 'Microsoft.Resources/deployments/write' over scope '/subscriptions/xxx Resources/deployments/azure-cog-seaf-ff1f36f91cbd/providers/Microsoft.Resources/deployments/azure-cog-search-dev-1693473006' or the scope is invalid. If access was recently granted, please refresh your credentials."
  }
}
--------------------------------------------------------------------------------**

**azd version: 1.2.0**

I have a few questions to ask regarding the necessity of write permissions. Specifically, what is the exact reason for requiring write permissions? I already have existing resources, such as GPT-4 and an embedding model. Would it still be necessary to grant write access?, and if i want to run locally in my pc, does it still require write access?

**Comment by pamelafox** [comment]
The Bicep files in the infra folder currently require that you have permissions to create a resource group and also assign various roles. If you do not have enough permissions to run `azd up`, then you could create each resource manually and set the required environment variables.


AZURE_FORMRECOGNIZER_RESOURCE_GROUP=""
AZURE_FORMRECOGNIZER_SERVICE=""
AZURE_LOCATION=""
AZURE_OPENAI_CHATGPT_DEPLOYMENT=""
AZURE_OPENAI_CHATGPT_MODEL=""
AZURE_OPENAI_EMB_DEPLOYMENT=""
AZURE_OPENAI_RESOURCE_GROUP=""
AZURE_OPENAI_SERVICE=""
AZURE_RESOURCE_GROUP=""
AZURE_SEARCH_INDEX=""
AZURE_SEARCH_SERVICE=""
AZURE_SEARCH_SERVICE_RESOURCE_GROUP=""
AZURE_STORAGE_ACCOUNT=""
AZURE_STORAGE_CONTAINER=""
AZURE_STORAGE_RESOURCE_GROUP=""
AZURE_SUBSCRIPTION_ID=""
AZURE_TENANT_ID=""

You can check app.py to see how these environment variables are used. Unfortunately quite a few are needed, and you also need prepdocs.py to be able to create the search index and data.

You might have more success using this other Chat on Your Data solution instead:

https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/use-your-data

That corresponds to this repo, but can be deployed entirely in the Portal:

https://github.com/microsoft/sample-app-aoai-chatGPT/

**Comment by pamelafox** [comment]
Closing this as a duplicate of https://github.com/Azure-Samples/azure-search-openai-demo/issues/4 but please let us know if you have further questions.

**Comment by Kunal1505** [comment]
Oh, hello @pamelafox , thank you for the explanation. I apologize for not realizing there was an existing duplicate issue.

I have a few questions to ask.

The first one is: when we use the 'azd up' command, which file gets executed first and how does the execution flow? Could you provide a brief overview of the process? I am familiar with how the backend works but I am interested in understanding this from a more programming-specific perspective.

My next question is, only after all the resources are deployed can I run it locally, correct? Or, is it the case that I need to deploy all resources and require write permissions only for the Azure web app?

**Comment by pamelafox** [comment]
@Kunal1505 I've added a description of `azd up` here:
https://github.com/Azure-Samples/azure-search-openai-demo/pull/602/files

Re free trial: $200 should be enough for an initial deploy. I think a recent colleague estimated theirs at $15/day, mostly in Cognitive Search. But let us know if you get permission issues around what you can provision on the free trial. You may need to modify the Bicep.

**Comment by pamelafox** [comment]
FYI: We've clarified on the README that at the minimum, you need the RBAC role to run `azd up`, or the RBAC role scoped to an existing resource group: https://github.com/Azure-Samples/azure-search-openai-demo#azure-account-requirements

We're interested to hear if your employers are able to give you that level of access (especially resource group scoped) or if that is still high of a level of access. 

I also documented in another issue how to manually assign the roles in the Portal, but I assume that will only work if you have RBAC access for that RG:

https://github.com/Azure-Samples/azure-search-openai-demo/issues/634#issuecomment-1721699823

If you cannot be granted RBAC at the resource group level, then I think the only other approach is to use keys instead of managed identity. That should be possible with changes to the credentials in app.py. We generally discourage keys when possible, however, so we haven't yet enabled that in the repo.
